indispensable-josef/src/commonMain/kotlin/at/asitplus/signum/indispensable/josef/JweDecrypted.kt

```diff
@@ -7,25 +7,25 @@ package at.asitplus.signum.indispensable.josef
  *
  * @see [JweEncrypted]
  */
-data class JweDecrypted(
+data class JweDecrypted<T: Any>(
     val header: JweHeader,
-    val payload: ByteArray
+    val payload: T,
 ) {
     override fun equals(other: Any?): Boolean {
         if (this === other) return true
         if (other == null || this::class != other::class) return false
 
-        other as JweDecrypted
+        other as JweDecrypted<*>
 
         if (header != other.header) return false
-        if (!payload.contentEquals(other.payload)) return false
+        if (!payload.equals(other.payload)) return false
```

Comment:
this needs discrimination: check for primitive arrays (sadly, each one separately), Array<*> and regular equals.

Reply:
it may be useful to have this globally somewhere:
```kotlin
inline infix fun <reified T> T.contentEqualsIfArray(other: T) = when(this) {
    is Array<*> -> (other is Array<*>) && this.contentEquals(other)
    is ByteArray -> (other is ByteArray) && this.contentEquals(other)
    is ShortArray -> (other is ShortArray) && this.contentEquals(other)
    is IntArray -> (other is IntArray) && this.contentEquals(other)
    is LongArray -> (other is LongArray) && this.contentEquals(other)
    is FloatArray -> (other is FloatArray) && this.contentEquals(other)
    is DoubleArray -> (other is DoubleArray) && this.contentEquals(other)
    is CharArray -> (other is CharArray) && this.contentEquals(other)
    is BooleanArray -> (other is BooleanArray) && this.contentEquals(other)
    else -> (this == other)
}

inline fun <reified T> T.contentHashCodeIfArray() = when(this) {
    is Array<*> -> this.contentHashCode()
    is ByteArray -> this.contentHashCode()
    is ShortArray -> this.contentHashCode()
    is IntArray -> this.contentHashCode()
    is LongArray -> this.contentHashCode()
    is FloatArray -> this.contentHashCode()
    is DoubleArray -> this.contentHashCode()
    is CharArray -> this.contentHashCode()
    is BooleanArray -> this.contentHashCode()
    else -> this.hashCode()
}

```

Reply:
this will be very useful, but do we put it into signum? on the one hand, it is not crypto related, on the other hand, this would be the easiest way to propagate it throughout all our projects. If we put it into another library, we should mos def scour for all sorts of helpers throughout our codebase and drag those along

Reply:
internally, not as part of the public api

Reply:
we do have a slimmed-down variant of this in asn1element as well. we almost certainly need in in the cose module and in VC-K